roachtest: include reads for graceful shutdown test

andrewbaptist opened the pull request:
Small changes to the test:
1) Increase the cluster size from 3 to 6.
2) Introduce both reads and follower reads.
3) Increase the length of the test run and the time between
	 starts/stops.
4) Include both reads and writes when measuring QPS.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112926)
<!-- Reviewable:end -->

andrewbaptist commented:
This isn't quite ready for review, because I need to figure out if I want to tie it to the change to reordering of nodes based on locality before latency.

 Also, I don't really like the tangledness of the injecting of the health function into DistSender. It isn't strictly necessary, but I didn't want to do a bigger refactor since we likely want to backport this change. If there is a cleaner way to do this without splitting up the NodeLiveneess code, I would definitely do that.

Commit:
roachtest: include reads for graceful shutdown test

Small changes to the test kv/gracefuldraining/nodes=3.
1) Increase the cluster size from 3 to 6.
2) Introduce both reads and follower reads.
3) Increase the length of the test run and the time between starts/stops.
4) Include both reads and writes when measuring QPS.

Epic: none

Release note: None

andrewbaptist commented:
@nvanbenschoten  I don't think https://github.com/cockroachdb/cockroach/pull/112926/files#diff-104a11712684e520f0affa95e6a053faeb713306a10f69ee99b9729cc9dd1996 is safe (changing the func pointer after construction). It maybe works, since there wouldn't have been any requests sent when this is constructed, but that is just "lucky". 

The alternative of refactoring the code to change the construction order is complex and not something I'd want to try and backport. Are there other good alternatives (other than a mutex around the use of this function every time)?

andrewbaptist commented:
Test running:
```
roachprod run $CLUSTER:7 "./cockroach workload run kv --read-percent 50 --follower-read-percent 50 {pgurl:1-5}"
```

Some graphics of this with and without the change:
Without the change (drain at ~18:01:30)
![image](https://github.com/cockroachdb/cockroach/assets/46975918/26f70219-3445-4520-994f-9445abf8e8e4)

With the change (drain at ~17:52:30)
![image](https://github.com/cockroachdb/cockroach/assets/46975918/a93a37ed-b1f3-437e-8e70-e94d2ba4d021)

andrewbaptist commented:
@nvanbenschoten This should be ready for review. I need to look at the failure in `TestBoundedStalenessDataDriven` but I think the problem is an incorrect assumption in the test. It had failed independently on the same line in a different PR during race, so I don't think this is a new failure.

andrewbaptist commented:
Hit #112048 - which conceivably could be related, but it doesn't appear to be causing that failure as we shouldn't have any health issues.

nvb reviewed (commented):
> Reviewed 4 of 11 files at r2, 1 of 4 files at r4, 7 of 7 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112926)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @michae2, @renatolabs, and @srosenberg)

___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 274 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NiaTvMrEKBOw_U-KJ8O:-NiaTvMrEKBOw_U-KJ8P:bk78bxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cf0488ebb268dee3af0b1cd5987e9e09a966bc/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L274)):*
> ```Go
> 	case ReplicaGCQueue:
> 		return nv.isAlive()
> 	case DistSender:
> ```

The way this turned out is quite nice.
___
*[`pkg/server/server.go` line 550 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NiaUc9K8NaHM_q56iFB:-NiaUc9K8NaHM_q56iFC:b9uzy95) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cf0488ebb268dee3af0b1cd5987e9e09a966bc/pkg/server/server.go#L550)):*
> ```Go
> 	// struct.
> 	distSender.SetHealthFunc(func(id roachpb.NodeID) bool {
> 		return nodeLiveness.GetNodeVitalityFromCache(id).IsLive(livenesspb.DistSender)
> ```

Calling `GetNodeVitalityFromCache` `O(repl x log(repl))` times per RPC feels like it could come with a sizable performance penalty. Internally, this is acquiring the liveness cache's read lock twice, grabbing an HLC clock reading, and checking the RPC connection health (which itself acquires a few mutexes in gossip and in the RPC context).

Have we measured how expensive a single one of these calls is in a realistic setting (e.g. on a 32-core machine with 10k qps)?

Do we have any other callers of this API on the data path?
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go` line 1268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NiaWmTrDMdcA9SADy-P:-NiaWmTrDMdcA9SADy-Q:b-s0egpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cf0488ebb268dee3af0b1cd5987e9e09a966bc/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go#L1268)):*
> ```Go
> 		Requests: b.Requests(),
> 	}
> 	br.Header.Timestamp = readTime
> ```

nit: can't you do `b.Header.Timestamp` for this, and then still use `db.Run`?


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
Looking at the perf data running:
```
./cockroach workload run kv --duration "1m" --read-percent 100 --follower-read-percent 100 --concurrency 200
```
WIth the setting disabled (no health check):
33375.0 ops/sec

With the setting enabled:
33445.9 ops/sec

However `OptimizeReplicaOrder` is much more expensive with it enabled:
Disabled - 0.33%
![image](https://github.com/cockroachdb/cockroach/assets/46975918/7f9f597d-7267-4174-a954-e2d002ba571b)
Enabled - 2.17%
![image](https://github.com/cockroachdb/cockroach/assets/46975918/839c3449-f3df-49de-8859-cb8b0e3ec9e1)

I'm going to try a few things to reduce this overhead, although I don't think it will go to 0. The biggest cost right now is recomputing it more than once per node (since it is done within sort) and also looking at the connection health.

andrewbaptist commented:
Closing this for now - the PR #114205 will temporarily supersede this.

andrewbaptist commented:
The only remaining part of this was the change to the roachtest to include follower reads in the testing. I'm changing to @kvoli as he has worked on this test before.

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r10, 25 of 25 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-:-NjZiQvVErMthj818uFO:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`-- commits` line 4 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NjZiFWO0lDafQG39vqN):*
nit: reference the name of the test, since it is changed here.
___
*[`-- commits` line 12 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NjZi9Z3-xARHZgEir8r):*
nit: rm empty line.
___
*[`pkg/cmd/roachtest/tests/kv.go` line 622 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NjZgjN5CjKE0o4ugJIi:-NjZgjN65fqLWhRFOS1B:bps1gq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f178aeb2d1a70d759275a01c2c6ef514c292967/pkg/cmd/roachtest/tests/kv.go#L622)):*
> ```Go
> 					case <-ctx.Done():
> 						return
> 					case <-time.After(2 * time.Minute):
> ```

Q: Why the wait increase?


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-:-NlyI19o6S_EJsNSbimW:b-814g8k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kvoli and @nvanbenschoten)

___
*[`-- commits` line 4 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NjZiFWO0lDafQG39vqN):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: reference the name of the test, since it is changed here.
</blockquote></details>

Done
___
*[`-- commits` line 12 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NjZi9Z3-xARHZgEir8r):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: rm empty line.
</blockquote></details>

Done
___
*[`pkg/cmd/roachtest/tests/kv.go` line 622 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NjZgjN5CjKE0o4ugJIi:-NlyGt1dAFuoJGZtg6mB:b-8nd7oz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f178aeb2d1a70d759275a01c2c6ef514c292967/pkg/cmd/roachtest/tests/kv.go#L622)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Q: Why the wait increase?
</blockquote></details>

This was to make sure we capture any "latent" impacts of the restart. Previously it was restarting before the system was fully healthy (not all caught up and leases rebalanced). A longer wait here prevents this.
___
*[`pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go` line 1268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NiaWmTrDMdcA9SADy-P:-NlyHB3r-OJqhgTfWZaA:b-u0i3i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cf0488ebb268dee3af0b1cd5987e9e09a966bc/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go#L1268)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: can't you do `b.Header.Timestamp` for this, and then still use `db.Run`?
</blockquote></details>

Thanks - changed.
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 274 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NiaTvMrEKBOw_U-KJ8O:-NlyHhjF-aOGXPz_lquL:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cf0488ebb268dee3af0b1cd5987e9e09a966bc/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L274)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The way this turned out is quite nice.
</blockquote></details>

:thumbsup: 
___
*[`pkg/server/server.go` line 550 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/112926#-NiaUc9K8NaHM_q56iFB:-NlyHmUa1btuBpmTkTmn:b-nm8gx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0cf0488ebb268dee3af0b1cd5987e9e09a966bc/pkg/server/server.go#L550)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Calling `GetNodeVitalityFromCache` `O(repl x log(repl))` times per RPC feels like it could come with a sizable performance penalty. Internally, this is acquiring the liveness cache's read lock twice, grabbing an HLC clock reading, and checking the RPC connection health (which itself acquires a few mutexes in gossip and in the RPC context).

Have we measured how expensive a single one of these calls is in a realistic setting (e.g. on a 32-core machine with 10k qps)?

Do we have any other callers of this API on the data path?
</blockquote></details>

This was pulled into the other PR - but now this is cached. It turned out the most expensive part of this was the `ConnHealth` which is done twice without the other changes.


<!-- Sent from Reviewable.io -->

andrewbaptist commented:
Note that this test will pass, but have long latency periods after each restart due to misrouted follower reads:
<img width="963" alt="image" src="https://github.com/cockroachdb/cockroach/assets/46975918/9d89de5c-072a-4264-9d6d-290a1ef633d4">

This is part of the work to fix 24.2, however since this test is currently asserting on throughput not latency we can leave it enabled with follower reads without a high risk of it flaking.

andrewbaptist commented:
bors r=kvoli
